recipe-client-addon/test/unit/test_PreferenceExperiments.js

```diff
@@ -0,0 +1,306 @@
+"use strict";
+
+Cu.import("resource://gre/modules/Preferences.jsm");
+Cu.import("resource://shield-recipe-client/lib/PreferenceExperiments.jsm");
+
+load("utils.js"); /* globals withMockPreferences */
+
+// Save ourselves some typing
+const {withMockExperiments} = PreferenceExperiments;
+
+function experimentFactory(attrs) {
+  return Object.assign({
+    name: "fakename",
+    branch: "fakebranch",
+    expired: false,
+    lastSeen: new Date().toJSON(),
+    preferenceName: "fake.preference",
+    preferenceValue: "falkevalue",
+    previousPreferenceValue: "oldfakevalue",
+  }, attrs);
+}
+
+// clearAllExperimentStorage
+add_task(withMockExperiments(function* (experiments) {
+  experiments["test"] = experimentFactory({name: "test"});
+  ok(yield PreferenceExperiments.has("test"), "Mock experiment is detected.");
+  yield PreferenceExperiments.clearAllExperimentStorage();
+  ok(
+    !(yield PreferenceExperiments.has("test")),
+    "clearAllExperimentStorage removed all stored experiments",
+  );
+}));
+
+// start should throw if an experiment with the given name already exists
+add_task(withMockExperiments(function* (experiments) {
+  experiments["test"] = experimentFactory({name: "test"});
+  yield Assert.rejects(
+    PreferenceExperiments.start("test", "branch", "fake.preference", "value"),
+    "start threw an error due to a conflicting experiment name",
+  );
+}));
+
+// start should throw if an experiment for the given preference is active
+add_task(withMockExperiments(function* (experiments) {
+  experiments["test"] = experimentFactory({name: "test", preferenceName: "fake.preference"});
+  yield Assert.rejects(
+    PreferenceExperiments.start("different", "branch", "fake.preference", "value"),
+    "start threw an error due to an active experiment for the given preference",
+  );
+}));
+
+// start should save experiment data, modify the preference, and register a
+// watcher.
+add_task(withMockExperiments(withMockPreferences(function* (experiments, mockPreferences) {
+  const startObserver = sinon.stub(PreferenceExperiments, "startObserver");
+  mockPreferences.set("fake.preference", "oldvalue");
+
+  yield PreferenceExperiments.start("test", "branch", "fake.preference", "newvalue");
+  ok("test" in experiments, "start saved the experiment");
+  ok(
+    startObserver.calledWith("test", "fake.preference", "newvalue"),
+    "start registered an observer",
+  );
+
+  const experiment = experiments["test"];
+  equal(experiment.name, "test", "start saved the experiment name");
+  equal(experiment.branch, "branch", "start saved the experiment branch");
+  equal(experiment.expired, false, "start saved the expired status");
+  equal(experiment.preferenceName, "fake.preference", "start saved the preference name");
+  equal(experiment.preferenceValue, "newvalue", "start saved the preference value");
+  equal(
+    experiment.previousPreferenceValue,
+    "oldvalue",
+    "start saved the previous preference value",
+  );
```

Comment:
Deep equal would compare the `lastSeen` date (or at least I thought it would), which I don't want for this test. I could extract the properties I do want to compare, though. Do you prefer that?